service/discovery/aws/storage.go

```diff
@@ -0,0 +1,382 @@
+	GetBucketLifecycleConfiguration(ctx context.Context,
+		params *s3.GetBucketLifecycleConfigurationInput,
+		optFns ...func(*s3.Options)) (*s3.GetBucketLifecycleConfigurationOutput, error)
+}
+
+// BucketPolicy matches the returned bucket policy in JSON from AWS
+type BucketPolicy struct {
+	ID        string
+	Version   string
+	Statement []Statement
+}
+type Statement struct {
+	Action   string
+	Effect   string
+	Resource []string
+	Condition
+}
+type Condition struct {
+	Bool
+}
+type Bool struct {
+	AwsSecureTransport bool `json:"aws:SecureTransport"`
+}
+
+// Name is the method implementation defined in the discovery.Discoverer interface
+func (d *awsS3Discovery) Name() string {
+	return "AWS Blob Storage"
+}
+
+// List is the method implementation defined in the discovery.Discoverer interface
+func (d *awsS3Discovery) List() (resources []voc.IsResource, err error) {
+	var encryptionAtRest voc.AtRestEncryption
+	var encryptionAtTransmit voc.TransportEncryption
+
+	log.Info("Starting List() in ", d.Name())
+	err = d.getBuckets()
+	if err != nil {
+		return
+	}
+	log.Println(d.buckets)
+	for _, bucket := range d.buckets {
+		log.Println("Getting resources for", bucket.name)
+		encryptionAtRest, err = d.getEncryptionAtRest(bucket.name)
+		if err != nil {
+			return
+		}
+		encryptionAtTransmit, err = d.getTransportEncryption(bucket.name)
+		if err != nil {
+			return
+		}
+		resources = append(resources, &voc.ObjectStorageResource{
+			StorageResource: voc.StorageResource{
+				Resource: voc.Resource{
+					ID:           bucket.arn,
+					Name:         bucket.name,
+					CreationTime: bucket.creationTime.Unix(),
+					Type:         []string{"ObjectStorage", "Storage", "Resource"},
+				},
+				AtRestEncryption: &encryptionAtRest,
+			},
+			HttpEndpoint: &voc.HttpEndpoint{
+				URL:                 bucket.endpoint,
+				TransportEncryption: &encryptionAtTransmit,
+			},
+		})
+	}
+	return
+}
+func (b bucket) String() string {
+	return fmt.Sprintf("[ARN: %v, Name: %v, Creation Time: %v, Number of objects: %v]", b.arn, b.name, b.creationTime, b.numberOfObjects)
+}
+
+// NewAwsStorageDiscovery constructs a new awsS3Discovery initializing the s3-client and isDiscovering with true
+func NewAwsStorageDiscovery(cfg aws.Config) discovery.Discoverer {
+	return &awsS3Discovery{
+		client:        s3.NewFromConfig(cfg),
+		buckets:       nil,
+		isDiscovering: true,
+	}
+}
+
+// S3ListBucketsAPI is the interface for the List function (used for mock testing)
+type S3ListBucketsAPI interface {
+	ListBuckets(ctx context.Context,
+		params *s3.ListBucketsInput,
+		optFns ...func(*s3.Options)) (*s3.ListBucketsOutput, error)
+}
+
+// getBuckets returns all buckets
+func (d *awsS3Discovery) getBuckets() (err error) {
+	log.Println("Getting buckets in s3...")
+	var resp *s3.ListBucketsOutput
+	resp, err = d.client.ListBuckets(context.TODO(), &s3.ListBucketsInput{})
+	if err != nil {
+		log.Error("Error occurred.")
```

Comment:
This is not a very good log line :) You can avoid logging the error alltogether here, this will be done by the calling function in the discovery service.

Reply:
I wasn't sure whether the if-statement in line 178 is always true or what is if it is not. If there is a case it is not true, then there would be at least a message that an error occurred and one could further investigate it.
But I could put a return statement withing the if block and then add this error log directly afterwards so that there will be an error message in any case (and yes, then also with the message content of err itself).
I wanted to log it here, since I want to keep the AWS stuff (like smithy.OperationError) out of the general discovery and keep it coupled to this package here (so there is no import of aws stuff in the general discoverer).

But, yes, we can just return the error and handle it in the caller.